rust/actyx/ax-aql/src/language/parser.rs

```diff
@@ -721,6 +722,100 @@ pub(crate) fn query_from_str(s: &str) -> Result<Query<'_>> {
     Ok(query)
 }
 
+pub(crate) struct QueryWorkflowAnalysisError<'a> {
+    pub(crate) span: Span<'a, Ident>,
+    pub(crate) error_type: QueryWorkflowAnalysisErrorType,
+}
+
+pub(crate) enum QueryWorkflowAnalysisErrorType {
+    UndeclaredEvent,
+    UndeclaredWorkflowCall,
+}
+
+pub(crate) struct QueryWorkflowAnalysis<'a> {
+    pub(crate) query: &'a Query<'a>,
+}
+
+impl<'a> QueryWorkflowAnalysis<'a> {
+    pub fn from(query: &'a Query) -> Self {
+        Self { query }
+    }
+
+    pub fn check(&'a self) -> Vec<QueryWorkflowAnalysisError> {
+        self.query
+            .workflows
+            .iter()
+            .flat_map(|(_ident, workflow)| self.check_workflow(workflow))
+            .collect()
+    }
+
+    fn check_workflow(&'a self, workflow: &'a Workflow) -> Vec<QueryWorkflowAnalysisError> {
+        self.check_steps(&workflow.steps)
+    }
+
+    fn check_steps(&'a self, steps: &'a NonEmptyVec<WorkflowStep<'a>>) -> Vec<QueryWorkflowAnalysisError> {
+        steps.iter().flat_map(|step| self.check_step(step)).collect()
+    }
+
+    fn check_step(&'a self, step: &'a WorkflowStep<'a>) -> Vec<QueryWorkflowAnalysisError> {
+        use super::workflow::WorkflowStep::*;
```

Comment:
Nitpick: Not a fan of using all the enum members. In this case is not particularly problematic but take a look into this https://youtu.be/8j_FbjiowvE?si=DOchI2_0vieFkEjg&t=103 (not the full video)

Reply:
wat??